internal/runbits/progressbar.go

```diff
@@ -0,0 +1,222 @@
+	if err != nil {
+		logging.Debug("Could not get terminal size, assuming width=120: %v", err)
+		tw = 120
+	}
+
+	// calculate the maximum width for a name displayed to the left of the progress bar
+	maxWidth := tw - 80 - 19 // 80 is the default size for the progressbar, 19 is taken by counters (up to 999) and percentage display
+	if maxWidth < 0 {
+		maxWidth = 4
+	}
+	if tw <= 105 && tw >= 40 {
+		maxWidth = 11 // enough space to spell "downloading"
+	}
+	return maxWidth
+}
+
+func (pb *progressBar) artifactState(id artifact.ArtifactID, step events.ArtifactSetupStep) *artifactStepBar {
+	steps, ok := pb.artifactStates[id]
+	if !ok {
+		steps = make(map[events.ArtifactSetupStep]*artifactStepBar)
+		pb.artifactStates[id] = steps
+	}
+	state, ok := steps[step]
+	if !ok {
+		state = &artifactStepBar{}
+		steps[step] = state
+	}
+	return state
+}
+
+func (pb *progressBar) BuildStarted(total int64) error {
+	if pb.buildBar == nil {
+		pb.buildBar = pb.addTotalBar("Building", total)
+	}
+	return nil
+}
+
+func (pb *progressBar) BuildIncrement() error {
+	if pb.buildBar == nil {
+		return errs.New("Build bar has not been initialized yet.")
+	}
+
+	pb.buildBar.Increment()
+	return nil
+}
+
+func (pb *progressBar) BuildCompleted(anyFailures bool) error {
+	if pb.buildBar == nil {
+		return errs.New("Build bar has not been initialized yet.")
+	}
+
+	// ensure that the build bar reports a completion event even if some builds have failed
+	if anyFailures {
+		pb.buildBar.Abort(false)
+	} else {
+		pb.buildBar.SetTotal(0, true)
+	}
+	return nil
+}
+
+func (pb *progressBar) InstallationStarted(total int64) error {
+	if pb.installBar == nil {
+		pb.installBar = pb.addTotalBar("Installing", total)
+	}
+	return nil
+}
+
+func (pb *progressBar) InstallationIncrement() error {
+	if pb.installBar == nil {
+		return errs.New("Installation bar has not been initialized yet.")
+	}
+
+	pb.installBar.Increment()
+	return nil
+}
+
+func (pb *progressBar) InstallationCompleted(anyFailures bool) error {
+	if pb.installBar == nil {
+		return errs.New("Installation bar has not been initialized yet.")
+	}
+
+	// ensure that the build bar reports a completion event even if some builds have failed
+	if anyFailures {
+		pb.installBar.Abort(false)
+	} else {
+		pb.installBar.SetTotal(0, true)
+	}
+	return nil
+}
+
+func (pb *progressBar) ArtifactStepStarted(artifactID artifact.ArtifactID, step events.ArtifactSetupStep, name string, total int64) error {
+	as := pb.artifactState(artifactID, step)
+	if as.bar != nil {
+		return errs.New("Progress bar can be initialized only once.")
+	}
+	as.bar = pb.addProgressBar(pb.buildStepName(step, name), total)
+	as.started = time.Now()
+
+	return nil
+}
+
+func (pb *progressBar) buildStepName(step events.ArtifactSetupStep, artifactName string) string {
+	var prefix string
+	switch step {
+	case events.Download:
+		prefix = "D.."
+	case events.Unpack:
+		prefix = ".U."
+	case events.Install:
+		prefix = "..I"
```

Comment:
I don't see why we need to be abbreviating stuff. I know you're working without a mock but try to maximize the user experience. This won't make sense to anyone but us.

Reply:
Fair point...